File: app/src/main/java/net/veldor/flibustaloader/view_models/MainViewModel.java
package net.veldor.flibustaloader.view_models;

import android.app.Application;
import android.util.DisplayMetrics;
import android.util.Log;
import android.util.SparseBooleanArray;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.work.Data;
import androidx.work.ExistingWorkPolicy;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkInfo;
import androidx.work.WorkManager;

import com.msopentech.thali.android.toronionproxy.AndroidOnionProxyManager;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.MyWebView;
import net.veldor.flibustaloader.database.entity.ReadedBooks;
import net.veldor.flibustaloader.interfaces.MyViewModelInterface;
import net.veldor.flibustaloader.selections.DownloadLink;
import net.veldor.flibustaloader.selections.FoundedBook;
import net.veldor.flibustaloader.updater.Updater;
import net.veldor.flibustaloader.utils.BookSharer;
import net.veldor.flibustaloader.utils.MyFileReader;
import net.veldor.flibustaloader.utils.MyPreferences;
import net.veldor.flibustaloader.utils.XMLHandler;
import net.veldor.flibustaloader.workers.AddBooksToDownloadQueueWorker;
import net.veldor.flibustaloader.workers.SearchWorker;

import java.util.ArrayList;
import java.util.Random;
import java.util.Stack;
import java.util.UUID;

public class MainViewModel extends GlobalViewModel implements MyViewModelInterface {
    private static final String ADD_TO_DOWNLOAD_QUEUE_ACTION = "add to download queue";
    public static final String MULTIPLY_DOWNLOAD = "multiply download";

    private final Stack<Integer> mClickedItemsStack = new Stack<>();

    public MainViewModel(@NonNull Application application) {
        super(application);
    }

    public int getViewMode() {
        return App.getInstance().getViewMode();
    }

    public void switchViewMode(int type) {
        App.getInstance().switchViewMode(type);
    }

    public LiveData<Boolean> startCheckUpdate() {
        return Updater.checkUpdate();
    }

    public void initializeUpdate() {
        Updater.update();
    }

    public void switchNightMode() {
        App.getInstance().switchNightMode();
    }

    public boolean getNightModeEnabled() {
        return App.getInstance().getNightMode();
    }

    public ArrayList<String> getSearchAutocomplete() {
        String content = MyFileReader.getSearchAutocomplete();
        return XMLHandler.getSearchAutocomplete(content);
    }

    public String getRandomBookUrl() {
        Random random = new Random();
        return App.BASE_BOOK_URL + random.nextInt(App.MAX_BOOK_NUMBER);
    }

    public void shareLink(MyWebView mWebView) {
        BookSharer.shareLink(mWebView.getUrl());
    }

    public void setBookRead(FoundedBook book) {
        // запущу рабочего, который отметит книгу как прочитанную
        ReadedBooks readedBook = new ReadedBooks();
        readedBook.bookId = book.id;
        App.getInstance().mDatabase.readedBooksDao().insert(readedBook);
    }

    public void clearHistory() {
        MyFileReader.clearAutocomplete();
    }

    public LiveData<WorkInfo> downloadSelected(SparseBooleanArray ids) {
        App.getInstance().mDownloadSelectedBooks = ids;
        OneTimeWorkRequest downloadSelected = new OneTimeWorkRequest.Builder(AddBooksToDownloadQueueWorker.class).addTag(ADD_TO_DOWNLOAD_QUEUE_ACTION).build();
        WorkManager.getInstance(App.getInstance()).enqueueUniqueWork(ADD_TO_DOWNLOAD_QUEUE_ACTION, ExistingWorkPolicy.REPLACE, downloadSelected);
        return WorkManager.getInstance(App.getInstance()).getWorkInfoByIdLiveData(downloadSelected.getId());
    }

    public LiveData<WorkInfo> downloadAll(boolean unloaded) {
        App.getInstance().mDownloadSelectedBooks = null;
        App.getInstance().mDownloadUnloaded = unloaded;
        OneTimeWorkRequest downloadSelected = new OneTimeWorkRequest.Builder(AddBooksToDownloadQueueWorker.class).addTag(ADD_TO_DOWNLOAD_QUEUE_ACTION).build();
        WorkManager.getInstance(App.getInstance()).enqueueUniqueWork(ADD_TO_DOWNLOAD_QUEUE_ACTION, ExistingWorkPolicy.REPLACE, downloadSelected);
        return WorkManager.getInstance(App.getInstance()).getWorkInfoByIdLiveData(downloadSelected.getId());
    }

    public void initiateMassDownload() {
        App.getInstance().initializeDownload();
    }

    public void cancelMassDownload() {
        WorkManager.getInstance(App.getInstance()).cancelAllWorkByTag(MULTIPLY_DOWNLOAD);
    }

    public Boolean checkDownloadQueue() {
        return App.getInstance().checkDownloadQueue();
    }

    public void addToDownloadQueue(DownloadLink downloadLink) {
        AddBooksToDownloadQueueWorker.addLink(downloadLink);
        if (MyPreferences.getInstance().isDownloadAutostart()){
            App.getInstance().initializeDownload();
        }
    }

    public UUID request(String s) {
        Log.d("surprise", "MainViewModel request 131: load " + s);
        Data inputData = new Data.Builder()
                .putString(SearchWorker.REQUEST, s)
                .build();
        // запущу рабочего, который выполнит запрос
        OneTimeWorkRequest searchWorkRequest = new OneTimeWorkRequest.Builder(SearchWorker.class).addTag(SearchWorker.WORK_TAG).setInputData(inputData).build();
        WorkManager.getInstance(App.getInstance()).enqueueUniqueWork(SearchWorker.WORK_TAG, ExistingWorkPolicy.REPLACE, searchWorkRequest);
        return searchWorkRequest.getId();
    }

    public void saveClickedIndex(int sClickedItemIndex) {
        mClickedItemsStack.push(sClickedItemIndex);
        Log.d("surprise", "MainViewModel saveClickedIndex 140: to stack added " + sClickedItemIndex + ", stack size is " + mClickedItemsStack.size());
    }

    public Integer getLastClickedElement() {
        if (!mClickedItemsStack.isEmpty()) {
            return mClickedItemsStack.pop();
        }
        return -1;
    }

    public int getHeight() {
        DisplayMetrics displayMetrics = App.getInstance().getResources().getDisplayMetrics();
        int dpHeight = (int) (displayMetrics.heightPixels / displayMetrics.density);
        return dpHeight - 50;
    }
}

File: app/src/main/java/net/veldor/flibustaloader/view_models/MainViewModel.kt
package net.veldor.flibustaloader.view_models

import android.app.Application
import net.veldor.flibustaloader.updater.Updater.checkUpdate
import net.veldor.flibustaloader.updater.Updater.update
import net.veldor.flibustaloader.utils.MyFileReader.getSearchAutocomplete
import net.veldor.flibustaloader.utils.XMLHandler.getSearchAutocomplete
import net.veldor.flibustaloader.utils.BookSharer.shareLink
import net.veldor.flibustaloader.utils.MyFileReader.clearAutocomplete
import net.veldor.flibustaloader.workers.AddBooksToDownloadQueueWorker.Companion.addLink
import net.veldor.flibustaloader.interfaces.MyViewModelInterface
import net.veldor.flibustaloader.App
import androidx.lifecycle.LiveData
import net.veldor.flibustaloader.MyWebView
import net.veldor.flibustaloader.selections.FoundedBook
import net.veldor.flibustaloader.database.entity.ReadedBooks
import android.util.SparseBooleanArray
import net.veldor.flibustaloader.workers.AddBooksToDownloadQueueWorker
import net.veldor.flibustaloader.selections.DownloadLink
import net.veldor.flibustaloader.workers.SearchWorker
import android.util.Log
import androidx.work.*
import net.veldor.flibustaloader.utils.PreferencesHandler
import java.util.*

class MainViewModel(application: Application) : GlobalViewModel(application), MyViewModelInterface {
    private val mClickedItemsStack = Stack<Int>()
    val viewMode: Int
        get() = PreferencesHandler.instance.viewMode

    fun switchViewMode(type: Int) {
        PreferencesHandler.instance.viewMode = type
    }

    fun startCheckUpdate(): LiveData<Boolean> {
        return checkUpdate()
    }

    fun initializeUpdate() {
        update()
    }

    fun switchNightMode() {
        PreferencesHandler.instance.nightMode = !PreferencesHandler.instance.nightMode
    }

    val nightModeEnabled: Boolean
        get() = PreferencesHandler.instance.nightMode
    val searchAutocomplete: ArrayList<String>
        get() {
            val content = getSearchAutocomplete()
            return getSearchAutocomplete(content)
        }
    val randomBookUrl: String
        get() {
            val random = Random()
            return App.BASE_BOOK_URL + random.nextInt(App.MAX_BOOK_NUMBER)
        }

    fun shareLink(mWebView: MyWebView) {
        shareLink(mWebView.url)
    }

    fun setBookRead(book: FoundedBook) {
        // запущу рабочего, который отметит книгу как прочитанную
        val readedBook = ReadedBooks()
        readedBook.bookId = book.id!!
        App.instance.mDatabase.readBooksDao().insert(readedBook)
    }

    fun clearHistory() {
        clearAutocomplete()
    }

    fun downloadSelected(ids: SparseBooleanArray?): LiveData<WorkInfo> {
        App.instance.mDownloadSelectedBooks = ids
        val downloadSelected = OneTimeWorkRequest.Builder(
            AddBooksToDownloadQueueWorker::class.java
        ).addTag(ADD_TO_DOWNLOAD_QUEUE_ACTION).build()
        WorkManager.getInstance(App.instance).enqueueUniqueWork(
            ADD_TO_DOWNLOAD_QUEUE_ACTION,
            ExistingWorkPolicy.REPLACE,
            downloadSelected
        )
        return WorkManager.getInstance(App.instance)
            .getWorkInfoByIdLiveData(downloadSelected.id)
    }

    fun downloadAll(unloaded: Boolean): LiveData<WorkInfo> {
        App.instance.mDownloadSelectedBooks = null
        App.instance.mDownloadUnloaded = unloaded
        val downloadSelected = OneTimeWorkRequest.Builder(
            AddBooksToDownloadQueueWorker::class.java
        ).addTag(ADD_TO_DOWNLOAD_QUEUE_ACTION).build()
        WorkManager.getInstance(App.instance).enqueueUniqueWork(
            ADD_TO_DOWNLOAD_QUEUE_ACTION,
            ExistingWorkPolicy.REPLACE,
            downloadSelected
        )
        return WorkManager.getInstance(App.instance)
            .getWorkInfoByIdLiveData(downloadSelected.id)
    }

    fun initiateMassDownload() {
        App.instance.initializeDownload()
    }

    fun cancelMassDownload() {
        WorkManager.getInstance(App.instance).cancelAllWorkByTag(MULTIPLY_DOWNLOAD)
    }

    fun checkDownloadQueue(): Boolean {
        return App.instance.checkDownloadQueue()
    }

    fun addToDownloadQueue(downloadLink: DownloadLink?) {
        addLink(downloadLink!!)
        if (PreferencesHandler.instance.isDownloadAutostart) {
            App.instance.initializeDownload()
        }
    }

    fun request(s: String): UUID {
        Log.d("surprise", "MainViewModel request 131: load $s")
        val inputData = Data.Builder()
            .putString(SearchWorker.REQUEST, s)
            .build()
        // запущу рабочего, который выполнит запрос
        val searchWorkRequest =
            OneTimeWorkRequest.Builder(SearchWorker::class.java).addTag(SearchWorker.WORK_TAG)
                .setInputData(inputData).build()
        WorkManager.getInstance(App.instance)
            .enqueueUniqueWork(SearchWorker.WORK_TAG, ExistingWorkPolicy.REPLACE, searchWorkRequest)
        return searchWorkRequest.id
    }

    fun saveClickedIndex(sClickedItemIndex: Int) {
        mClickedItemsStack.push(sClickedItemIndex)
        Log.d(
            "surprise",
            "MainViewModel saveClickedIndex 140: to stack added " + sClickedItemIndex + ", stack size is " + mClickedItemsStack.size
        )
    }

    val lastClickedElement: Int
        get() = if (!mClickedItemsStack.isEmpty()) {
            mClickedItemsStack.pop()
        } else -1
    val height: Int
        get() {
            val displayMetrics = App.instance.resources.displayMetrics
            val dpHeight = (displayMetrics.heightPixels / displayMetrics.density).toInt()
            return dpHeight - 50
        }

    companion object {
        private const val ADD_TO_DOWNLOAD_QUEUE_ACTION = "add to download queue"
        const val MULTIPLY_DOWNLOAD = "multiply download"
    }
}